unittests/test_schedulers.py

```diff
@@ -544,116 +579,257 @@ def setUp(self):
+        self.assertEqual(self.testjob.num_tasks, 4)
+
+    def test_valid_constraint_partition(self):
         self.testjob.options = ['-C f1 f2', '--partition=p1 p2']
-        if expected_num_tasks:
+        self.prepare_job()
+        self.assertEqual(self.testjob.num_tasks, 4)
+
+    def test_not_valid_partition_cmd(self):
+        self.testjob._sched_partition = 'invalid'
+        with self.assertRaises(JobError):
             self.prepare_job()
-            self.assertEqual(self.testjob.num_tasks, expected_num_tasks)
-        else:
-            self.assertRaises(JobError, self.prepare_job)
 
-    def test_not_valid_partition(self):
-        self.testjob._sched_reservation = 'Foo'
-        self.testjob._sched_partition = 'Invalid'
-        self.assertRaises(JobError, self.prepare_job)
+    def test_not_valid_partition_short(self):
+        self.testjob.options = ['-p invalid']
+        with self.assertRaises(JobError):
+            self.prepare_job()
+
+    def test_not_valid_partition_long(self):
+        self.testjob.options = ['--partition=invalid']
+        with self.assertRaises(JobError):
+            self.prepare_job()
 
     def test_not_valid_constraint(self):
-        self.testjob._sched_reservation = 'Foo'
         self.testjob.options = ['--constraint=invalid']
-        self.assertRaises(JobError, self.prepare_job)
+        with self.assertRaises(JobError):
+            self.prepare_job()
+
+    def test_valid_reservation_cmd(self):
+        self.testjob._sched_access = ['--constraint=f2']
+        self.testjob._sched_reservation = 'dummy'
+        self.testjob._get_reservation_nodes = self.create_reservation_nodes
+        self.prepare_job()
+        self.assertEqual(self.testjob.num_tasks, 4)
 
-    def test_noreservation(self):
-        self.assertRaises(JobError, self.prepare_job)
+    def test_valid_reservation_option(self):
+        self.testjob._sched_access = ['--constraint=f2']
+        self.testjob.options = ['--reservation=dummy']
+        self.testjob._get_reservation_nodes = self.create_reservation_nodes
+        self.prepare_job()
+        self.assertEqual(self.testjob.num_tasks, 4)
+
+    def test_exclude_nodes_cmd(self):
+        self.testjob._sched_access = ['--constraint=f1']
+        self.testjob._sched_exclude_nodelist = 'nid00001'
+        self.testjob._get_nodes_by_name = self.get_nodes_by_name
+        self.prepare_job()
+        self.assertEqual(self.testjob.num_tasks, 8)
+
+    def test_exclude_nodes_short(self):
+        self.testjob._sched_access = ['--constraint=f1']
+        self.testjob.options = ['-x nid00001']
+        self.testjob._get_nodes_by_name = self.get_nodes_by_name
+        self.prepare_job()
+        self.assertEqual(self.testjob.num_tasks, 8)
+
+    def test_exclude_nodes_long(self):
+        self.testjob._sched_access = ['--constraint=f1']
+        self.testjob.options = ['--exclude=nid00001']
+        self.testjob._get_nodes_by_name = self.get_nodes_by_name
+        self.prepare_job()
```

Comment:
As a general comment about the unit tests, I think that testing the short and long options separately is a bit too much. There would be no such need if you follow my suggestion and use `argparse` also for parsing these options.

Reply:
Applied the suggested change